src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
+                    NullCount--;
+                }
+            }
+            else
+            {
+                if ((curBitMap >> ((int)(index % 8)) & 1) == 1 && index < Length)
+                { 
+                    // old value was NOT null and new value is null
+                    NullCount++;
+                }
+                else if (index == Length)
+                {
+                    // New entry from an append
+                    NullCount++;
+                }
+                newBitMap = (byte)(curBitMap & (byte)~(1 << (int)(index % 8)));
+            }
+            bitMapBuffer[bitMapBufferIndex] = newBitMap;
+        }
+
+        private bool GetBit(long index)
+        {
+            if (index < 0 || index > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Get the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
+            Debug.Assert(bitMapBuffer.Length > bitMapBufferIndex);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            return ((curBitMap >> ((int)index % 8)) & 1) != 0;
+        }
+
         public long Length;
         //TODO:
-        public long NullCount => throw new NotImplementedException();
+        public long NullCount;
         private int GetArrayContainingRowIndex(ref long rowIndex)
         {
             if (rowIndex > Length)
             {
                 throw new ArgumentOutOfRangeException(Strings.ColumnIndexOutOfRange, nameof(rowIndex));
             }
-            int curArrayIndex = 0;
-            int numBuffers = Buffers.Count;
-            while (curArrayIndex < numBuffers && rowIndex > Buffers[curArrayIndex].Length)
-            {
-                rowIndex -= Buffers[curArrayIndex].Length;
-                curArrayIndex++;
-            }
-            return curArrayIndex;
+            return (int)(rowIndex / Buffers[0].MaxCapacity);
         }
 
         public IList<T> this[long startIndex, int length]
         {
             get
             {
                 var ret = new List<T>(length);
-                int arrayIndex = GetArrayContainingRowIndex(ref startIndex);
-                bool temp = Buffers[arrayIndex][(int)startIndex, length, ret];
-                while (ret.Count < length)
+                long endIndex = Math.Min(Length, startIndex + length);
+                for (long i = startIndex; i < endIndex; i++)
                 {
-                    arrayIndex++;
-                    temp = Buffers[arrayIndex][0, length - ret.Count, ret];
+                    T? value = this[i];
+                    if (value.HasValue)
```

Comment:
Is this correct - to skip any nulls? Would the caller expect that passing in a `length` of `10` be returned with a list of `8` values?

Reply:
Yup, this was wrong. Fixed it